### ISSUE TITLE
[BUG] Get-PnPUser -WithRightsAssigned times out while working with some sites

### ISSUE THREAD

**Issue by Nazul** [pnp/powershell#3162]
### Expected behavior
`Get-PnPUser -WithRightsAssigned` should return the list of users with rights assigned.

### Actual behavior 
`Get-PnPUser -WithRightsAssigned` times out.
![image](https://github.com/pnp/powershell/assets/6336180/f1b45f0e-e0a9-4034-b2d9-0ebaf7c8a7f1)

### Steps to reproduce behavior
1. Connect to a SharePoint site
2. Run `Get-PnPUser -WithRightsAssigned`, which fails

Alternate path
1. Connect to a SharePoint site
2. Get current context and change timeout value
3. `Get-PnPUser -WithRightsAssigned` stills fails after running for about 1:53-1:55 minutes

### What is the version of the Cmdlet module you are running?
This was tested with 1.12.0 (Windows PowerShell 5.1) as well as with 2.1.1 (PowerShell 7.3.4)

### Which operating system/environment are you running PnP PowerShell on?
- [X] Windows
- [ ] Linux
- [ ] MacOS
- [ ] Azure Cloud Shell
- [ ] Azure Functions
- [ ] Other : please specify

**Comment by veronicageek** [comment]
@Nazul - Can you try on another site? I have no issue on my side (using `2.1.69` nightly)

![image](https://github.com/pnp/powershell/assets/25181757/723e8b12-03a6-4416-a8b7-692e4b5f5c93)

**Comment by Nazul** [comment]
I thought that I mentioned. It doesn't fail with all sites, but just with some. For those were it fails, it fails with both 1.1.12 and 2.1.1 versions.

**Comment by veronicageek** [comment]
In that case it might be more of an issue with the backend, and there are currently some issues with M365 (details in the tenant admin center). If it works even for some sites, not sure we can do anything unfortunately.

**Comment by Nazul** [comment]
We are aware of the Microsoft services outage, however, the issue we described occurred before the outage according to our logs, and it is still there even though SharePoint is listed as fully operational today for our tenant.

The problem that I see related with `Get-PnPUser` is that we have no way to modify the timeout value, so we can wait longer for the operation to be completed.

**Comment by veronicageek** [comment]
Well, I'll leave @gautamdsheth or @KoenZomers look at it then.

**Comment by patritasan** [comment]
I'm facing same issue. In my case I'm using PowerShell 7.3.7 and PnP.PowerShell 2.2.0
Do we have any news about possible solution?

**Comment by danny-yehya** [comment]
While I am not getting the specific timeout error directly, we are getting "Get-PnPUser: Maximum retry attempts 10, has be attempted." whenever this is run on the base site https://_tenant_.sharepoint.com . This tenant has about 45K active users, and the total user count based on some other things run shows approximately 87K users.

This needs to be able to handle throttling and retries better. I did not go through all the code to see how it is internally managing but with a large active tenant, the code should be able to respect throttling from SharePoint and retry based on that.

**Comment by vishnookumar** [comment]
Any progress made on this issue? I noticed that both errors: "The Operation has timed out & Maximum retry attempts 10, has be attempted" are reported only on sites which has more than 10k active users. Currently there is no way to batch or do pagination for this cmdlet: Get-PnPUser. Can this be fixed please?

**Comment by IfEyeOnlyKnew** [comment]
Seeing the same issue and yes over 10K users, "Maximum retry attempts 10, has be attempted."
"Currently there is no way to batch or do pagination for this cmdlet: Get-PnPUser. Can this be fixed please?"

**Comment by ryszardcopa** [comment]
I have the same issue.
Get-PnPUser -WithRightsAssigned -Connection $connection
Get-PnPUser : Maximum retry attempts 10, has be attempted.

**Comment by SriramG04** [comment]
I am facing the same issue. Does it timeout for the sites that have large number of users?

**Comment by reshmee011** [comment]
@SriramG04, could you confirm the number of users on the sites where you’re experiencing timeouts compared with the sites where everything is working normally?

**Comment by reshmee011** [comment]
@Nazul : I’m closing this issue due to no response for nearly 1 year. Please feel free to open a new issue if you’d like to revisit it.

**Comment by Nazul** [comment]
> @Nazul : I’m closing this issue due to no response for nearly 1 year. Please feel free to open a new issue if you’d like to revisit it.

@reshmee011 yes, there's no response from the PnP team so I already assumed this was never going to be fixed and moved on to different approach.

**Comment by reshmee011** [comment]
@Nazul It’s difficult to reproduce issues that occur only intermittently. Please share the alternative approach you used, as it may help others facing similar problems and struggling to find a workaround. In some cases, the underlying API returning the data may be experiencing throttling. Out of curiosity how many users were in your site where the issue was happening.

**Comment by reshmee011** [comment]
If you are looking at user permissions, you may consider the following script: https://pnp.github.io/script-samples/spo-get-permission-audit/README.html?tabs=pnpps. The get-pnpuser is using the CSOM API which probably is taking a long time to come back if there are too many unique permissions.

**Comment by Nazul** [comment]
@reshmee011 when we had the issue and I reported it, it was not intermittent. I mentioned that it happened with some sites (small sites) while it failed with others (big sites). Always the same sites.

In my opinion, it's just a matter of cmdlet was unable to finish because of the data size and it got terminated. A timeout parameter could help.

Anyway, we are no longer relying on this so it is up to your team to decide if they close this or not.

We have no viable alternative. We simple decide not to use this data for the process we were implementing.

Finally, thank you for the link. I'll review it and see if it works for future processes we could implement.

**Comment by gautamdsheth** [comment]
Hello, has been a long time. I took some time today and added some fixes which will be available starting with tomorrow's nightly builds. Could you test it if possible and let us know ? Added some server side filtering and pagination specific to this parameter , so should help.
